src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2146,8 +2204,22 @@ install_deployment_kernel (OstreeSysroot *sysroot, int new_bootversion,
   if (!glnx_opendirat (sysroot->boot_fd, bootconfdir, TRUE, &bootconf_dfd, error))
     return FALSE;
 
+  g_autofree char *bootconf_filename;
+  if (bootloader_is_boot_count_enabled(sysroot, cancellable, error))
+    {
+      guint max_tries = bootloader_get_max_boot_tries (sysroot, cancellable, error);
+      bootconf_filename = g_strdup_printf ("%s+%u.conf", bootconf_name, max_tries);
+
+      if (!bootloader_remove_tmp_entries(bootconf_dfd, bootconf_name, max_tries, cancellable, error))
+        return FALSE;
+    }
+  else
+    {
+      bootconf_filename = g_strdup_printf ("%s.conf", bootconf_name);
+    }
+
   if (!ostree_bootconfig_parser_write_at (ostree_deployment_get_bootconfig (deployment),
-                                          bootconf_dfd, bootconf_name, cancellable, error))
+                                          bootconf_dfd, bootconf_filename, cancellable, error))
```

Comment:
Hmm, but note that this function is called for _every_ deployment on _every_ update (that results in a BLS write). So this would effectively "unbless" a boot that was previously blessed (or reset to the max boots that did have some failures).

ISTM like the boot count needs to be part of e.g. the `OstreeDeployment` object so that we can leave it untouched when rewriting BLS entries.

Reply:
Technically we can have just a boolean flag (`bootable` field for example) for that, and check if a filename of an old BLS entry (stored in `/boot/loader.%d/entries`) still has a boot counter suffix. And if it doesn't -> set OstreeDeployment->bootable, so then we can create all new BLS entries for this particular deployment without any boot counters.

The problem is that deployment index for OstreeDeployment is not static and there is no deterministic way to match existing boot entry with a specific OstreeDeployment, maybe only parsing BLS entry contents could help

Reply:
> The problem is that deployment index for OstreeDeployment is not static and there is no deterministic way to match existing boot entry with a specific OstreeDeployment, maybe only parsing BLS entry contents could help

There is, but that state is currently not kept around. If you look at e.g. `sysroot_load_from_bootloader_configs()` is where we go through the BLS entries and convert those to an ordered list of `OstreeDeployment`s. So we could parse the counter information from the filename there and keep that state in the deployment objects.

Reply:
Re-implemented as you suggested, please take a look